Adds interface to send ECN bits for WebRTC from 'Network Process' to 'Web Process'.

xingri opened the pull request:
#### e243885913ef60bed4b13c7efce0abf0a17f2541
<pre>
Adds interface to send ECN bits for WebRTC from &apos;Network Process&apos; to &apos;Web Process&apos;.

<a href="https://bugs.webkit.org/show_bug.cgi?id=271191">https://bugs.webkit.org/show_bug.cgi?id=271191</a>

Reviewed by Youenn Fablet.

ECN means Explicit Congestion Notification.
See RFC9330 for details.

* Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:
(WebKit::LibWebRTCSocketClient::signalReadPacket):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::NetworkRTCTCPSocketCocoa):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::getECN):
(WebKit::processUDPData):
(WebKit::NetworkRTCUDPSocketCocoaConnections::setupNWConnection):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RTCNetwork.h:
* Source/WebKit/Shared/RTCNetwork.serialization.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
(WebKit::convertToWebRTCEcnMarking):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.messages.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.cpp:
(WebKit::LibWebRTCSocket::signalReadPacket):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.h:
Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:

Canonical link: <a href="https://commits.webkit.org/282674@main">https://commits.webkit.org/282674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4599874a5587858a2e0ab0dbd79990cbb7faa68

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63847 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43204 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67869 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14455 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14735 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51450 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10000 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40022 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55276 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32066 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/63359 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36702 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13328 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58671 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69565 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7794 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12528 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58772 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55372 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58936 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14136 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6483 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39024 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40103 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39846 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (commented):
Let's use RTCNetwork.serialization.in to allow IPC to directly send the etc::EcnMarking enum.

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -129,6 +129,41 @@ static dispatch_queue_t udpSocketQueue()
+    rtc::EcnMarking ecn = rtc::EcnMarking::kNotEct;
+
+    nw_protocol_definition_t protocol = nw_protocol_copy_ip_definition();
+    nw_protocol_metadata_t metadata = nw_content_context_copy_protocol_metadata(nwContext, protocol);
```
I would tend to use auto for both, and use adoptNS/adoptCF for metadata, so as to not call nw_release.
In its current form, if nw_protocol_metadata_is_ip returns false, `metadata` would leak.

xingri replied:
@youennf  thanks for the review. updated.

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -129,6 +129,40 @@ static dispatch_queue_t udpSocketQueue()
+        nw_ip_ecn_flag_t ecnFlag = nw_ip_metadata_get_ecn_flag(metadata.get());
+        switch (ecnFlag) {
+        case nw_ip_ecn_flag_non_ect:
+            ecn = RTC::Network::EcnMarking::kNotEct;
```
It would read better as:
`return RTC::Network::EcnMarking::kNotEct;` here and below, even for the default case.

xingri replied:
@youennf  Thanks for the review. Updated.

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -129,6 +129,40 @@ static dispatch_queue_t udpSocketQueue()
+        default:
+            break;
+        }
+    } else {
```
If we return for all cases, we can remove this else and do:
```
RELEASE_LOG_IF(!metadata, WebRTC, "Could not retreive the metadata from UDPSocket Context, so use default ECN value");
return RTC::Network::EcnMarking::kNotEct;
```

youennf commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
```diff
@@ -111,18 +111,33 @@ void LibWebRTCNetwork::dispatch(Function<void()>&& callback)
+        return rtc::EcnMarking::kEct0;
+    case RTC::Network::EcnMarking::kCe:
+        return rtc::EcnMarking::kCe;
+    }
```
Let's add `ASSERT_NOT_REACHED();`

xingri replied:
@youennf  Thanks for the review. Updated.

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -129,6 +129,40 @@ static dispatch_queue_t udpSocketQueue()
+    auto metadata = adoptNS(nw_content_context_copy_protocol_metadata(nwContext, protocol));
+
+    if (metadata && nw_protocol_metadata_is_ip(metadata.get())) {
+        nw_ip_ecn_flag_t ecnFlag = nw_ip_metadata_get_ecn_flag(metadata.get());
```
auto is good here too.

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -129,6 +129,40 @@ static dispatch_queue_t udpSocketQueue()
+    // Set Default ECN value
+    RTC::Network::EcnMarking ecn = RTC::Network::EcnMarking::kNotEct;
+
+    auto protocol = nw_protocol_copy_ip_definition();
```
adoptNS here as well I think.

youennf reviewed (approved):
LGTM with a few suggestions below.

xingri committed:
Adds interface to send ECN bits for WebRTC from 'Network Process' to 'Web Process'.

https://bugs.webkit.org/show_bug.cgi?id=271191

Reviewed by Youenn Fablet.

ECN means Explicit Congestion Notification.
See RFC9330 for details.

* Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:
(WebKit::LibWebRTCSocketClient::signalReadPacket):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::NetworkRTCTCPSocketCocoa):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::getECN):
(WebKit::processUDPData):
(WebKit::NetworkRTCUDPSocketCocoaConnections::setupNWConnection):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RTCNetwork.h:
* Source/WebKit/Shared/RTCNetwork.serialization.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
(WebKit::convertToWebRTCEcnMarking):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.messages.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.cpp:
(WebKit::LibWebRTCSocket::signalReadPacket):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.h:
Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:

Canonical link: https://commits.webkit.org/282674@main
